File: edge-vision-uav-landing/src/control_py/pid_controller.py
class PIDController:
    def __init__(self, kp, ki, kd, v_max, deadband=0.05, alpha=0.5):

        self.kp = kp
        self.ki = ki
        self.kd = kd
        self.v_max = v_max
        self.deadband = deadband
        self.alpha = alpha
        self.reset()

    def reset(self):
        """Gọi khi target loss để xóa memory, tránh vọt lố khi bắt lại mục tiêu."""
        self.integral = 0.0
        self.prev_error = 0.0
        self.prev_derivative = 0.0
        self.is_first_run = True

    def compute(self, error, dt):
        if dt <= 0.0:
            return 0.0

        # Deadband policy
        if abs(error) < self.deadband:
            error = 0.0

        p_term = self.kp * error

        # First-sample derivative handling
        if self.is_first_run:
            raw_derivative = 0.0
            self.is_first_run = False
        else:
            raw_derivative = (error - self.prev_error) / dt

        # Low-pass filter cho đạo hàm
        derivative = self.alpha * raw_derivative + (1 - self.alpha) * self.prev_derivative
        d_term = self.kd * derivative

        # Anti-windup (Clamping với unwind support)
        pre_cmd = p_term + d_term + self.ki * (self.integral + error * dt)
        
        # Chỉ cộng dồn I nếu hệ thống chưa bị bão hòa, 
        # HOẶC nếu lỗi đang có xu hướng kéo hệ thống ra khỏi bão hòa (đảo chiều)
        if abs(pre_cmd) < self.v_max or (pre_cmd > 0 and error < 0) or (pre_cmd < 0 and error > 0):
            self.integral += error * dt

        i_term = self.ki * self.integral

        cmd = p_term + i_term + d_term
        
        # Saturation
        cmd_clamped = max(-self.v_max, min(self.v_max, cmd))

        self.prev_error = error
        self.prev_derivative = derivative

        return cmd_clamped

File: edge-vision-uav-landing/src/control_cpp/src/pid_controller.cpp
#include "pid_controller.hpp"
#include <cmath>
#include <algorithm>

namespace control {

PIDController::PIDController(double kp, double ki, double kd, double v_max, double deadband, double alpha) 
    : kp_(kp), ki_(ki), kd_(kd), v_max_(v_max), deadband_(deadband), alpha_(alpha) {
    reset();
}

void PIDController::reset() {
    integral_ = 0.0;
    prev_error_ = 0.0;
    prev_derivative_ = 0.0;
    is_first_run_ = true;
}

double PIDController::compute(double error, double dt) {
    if (dt <= 0.0) {
        return 0.0;
    }

    if (std::abs(error) < deadband_) {
        error = 0.0;
    }

    double p_term = kp_ * error;

    double raw_derivative = 0.0;
    if (is_first_run_) {
        raw_derivative = 0.0;
        is_first_run_ = false;
    } else {
        raw_derivative = (error - prev_error_) / dt;
    }

    double derivative = alpha_ * raw_derivative + (1.0 - alpha_) * prev_derivative_;
    double d_term = kd_ * derivative;

    double pre_cmd = p_term + d_term + ki_ * (integral_ + error * dt);

    if (std::abs(pre_cmd) < v_max_ || (pre_cmd > 0 && error < 0) || (pre_cmd < 0 && error > 0)) {
        integral_ += error * dt;
    }

    double i_term = ki_ * integral_;
    double cmd = p_term + i_term + d_term;

    double cmd_clamped = std::max(-v_max_, std::min(v_max_, cmd));

    prev_error_ = error;
    prev_derivative_ = derivative;

    return cmd_clamped;
}

} // namespace control
